### DISCUSSION TITLE
Is there any documentation on how to write tests?

### DISCUSSION THREAD

**Discussion by cjbland** [remix-run/remix#993] [Q&A] [answered]
I'm trying to write tests using jest and am running into some issues. I tried writing a test that renders the main blog example page and got an error: `Error: You must render this element in a remix route element`. Here's what I did:

```bash
cd blog-tutorial
npm install
npm install -D @babel/preset-env @babel/preset-typescript @testing-library/jest-dom @testing-library/react @testing-library/user-event @types/jest babel-jest jest ts-jest
vim babel.config.js
vim jest.config.js
vim jest.setup.tsx
cd app/routes/
mkdir __tests__
cd __tests__
vim Main.test.tsx
```

And my files look as follows:

babel.config.js
```javascript
module.exports = {
  presets: [
    ['@babel/preset-env', {targets: {node: 'current'}}],
    '@babel/preset-typescript',
  ],
};
```

jest.config.js
```javascript
module.exports = { 
  preset: 'ts-jest',
  setupFilesAfterEnv: [`<rootDir>/jest.setup.tsx`],
  testEnvironment: 'jsdom',
  testPathIgnorePatterns: [`node_modules`, `\\.cache`, `<rootDir>.*/public`],
  testURL: `http://localhost`,
  transform: {
    '^.+\\.(ts|tsx)?$': 'ts-jest',
    '^.+\\.(js|jsx)$': 'babel-jest'
  }
};

```

jest.setup.tsx
```typescript
import "@testing-library/jest-dom/extend-expect";
```

Main.test.tsx
```typescript
import Main from "../index";
import { render, screen } from "@testing-library/react";

describe("Main::", () => {
  test("it works", () => {
    render(<Main />);
    screen.debug();
  }); 
});
```

Looking at the `entry.client.tsx` and `entry.server.tsx` files I'm assuming a `RemixBrowser` needs to be added in there somewhere, but it's not exactly clear to me how to do that. Are there any suggestions on how to properly write tests for a Remix app? Or is there a write-up somewhere that goes over these things?

**Comment by kvnwolf** [comment]
I think, for testing this kind of project, you should rely more on e2e tests rather than component tests. The kentcdodds.com source is always a good inspiration to me: https://github.com/kentcdodds/kentcdodds.com

**Reply by eran-or** [reply]
There's no real example of where to place the tests and how to test a components in remix context. There's only examples of e2e testing and a __testing_routes__ folder with js file which is not even a .test. file.

**Reply by cjbland** [reply]
What I ended up doing, and I'm not 100% happy with it, but it works, is I created a `components` folder as a sibling to the `routes` folder. The code in my individual routes is relatively small as they instantiate an object from `components`. I'm able to write proper unit tests for each of my components and then I use cypress to handle the integration tests. My unit test code coverage excludes the `routes` folder just to make me feel like I have better coverage :) I can provide my configs and some samples if you'd like.

**Comment by sergiodxa** [comment, marked as answer]
I recommend you to
1. Unit tests loaders and action (see https://sergiodxa.com/articles/test-remix-loaders-and-actions)
2. Test with RTL any component not depending on Remix hooks or components (they need a context that you don’t have access to work)
3. Use Cypress to do E2E tests to test whole routes

Also check #372

**Reply by phollome** [reply]
Thank you for your recommendation. As a hint for people reading your article and using Jest. You have to add
```js
import { installGlobals } from "@remix-run/node";
installGlobals();
```
to your Jest setup file.

Fixes the `Request is not defined` issue.

**Reply by cjbland** [reply]
Which version of jest are you using? I recently upgraded from v27 to v28 and now I'm getting this error:

> TypeError: Class extends value undefined is not a constructor or null
>  \> 1 | import { installGlobals } from "@remix-run/node/globals";

Did you happen to come across this? I've been trying to track this down for a few hours but no luck. I did have to add this line to my `jest.config.js` file after the upgrade too:

```
{
  ...
  transformIgnorePatterns: [
    "/node_modules/@remix-run/(?!web-blob|web-fetch|web-file|web-form-data|web-stream)",
  ],
}
```

**Reply by cjbland** [reply]
For those keeping score, I tried upgrading to jest 28 again today and ran into this same problem with `installGlobals`. I'll document anything I come across that gets the tests to go through.

**Reply by phollome** [reply]
Were you able to solve your problems after the update? I've updated to jest 28 and it works. I had to add `jest-environment-jsdom` and `@web-std` to `transformIgnorePatterns`.

**Reply by cjbland** [reply]
No, I was not. My `transformIgnorePatterns` ended up getting really large which made me think I was doing something wrong. In the end I converted them to use vitest since that's what the Remix Stacks use.

**Reply by MoSattler** [reply]
Hey @sergiodxa, unfortunately this doesn't seem to work (and the examples of your blog seems outdated). 
I tried it with the blues stack.

- `new Request()` needs an absolute URL, not a relative as in your examples. 
- it doesn't work when using things like `redirect`. It gives an odd error: `Unknown Error: undefined`

**Reply by sergiodxa** [reply]
> new Request() needs an absolute URL, not a relative as in your examples

Yeah this used to work, a few releases ago Remix created a polyfill for fetch/Request/Response objects and now they need an absolute URL.

> it doesn't work when using things like redirect. It gives an odd error: Unknown Error: undefined

I'm not sure why it would throw that

**Reply by MoSattler** [reply]
> I'm not sure why it would throw that

this seems to happen with a freshly setup remix app - so I guess it's some implementation detail of remix?

**Reply by MoSattler** [reply]
Never mind, it was actually throwing a redirect, so an error was expected. Just the `Unknown Error: undefined` that vitest gave out was confusing.

**Comment by cjbland** [comment]
Thanks for the Cypress tip @sergiodxa that definitely has helped me. When I figure out a solid pattern I'll update this discussion for someone else who may be trying to do the same things. So far using Cypress like I normally would with a `create-react-app` works just fine. It's dealing with the server interactions that I'm having the most trouble with (i.e. spying/intercepting function calls to third party libraries).

**Comment by marksmall** [comment]
Thanks for the response to the orignators question I will certainly use `cypress`, neither am I trying to rock the boat or anything, but the answer doesn't seem to be very realistic. 

Writing unit testing and using `Storybook` for developing components in isolation give more confidence in components we write and can make it easier and more productive, developing components. On a project of any real scale, there are going to be many components that rely on router hooks e.g. useParams, useLocation etc. 

Maybe I'm just using remix wrong, that is a very real possibility, if so, please say. Take Storybook as an example, most projects I've worked on have found this to be really useful in developing components, if I can't wrap a component in a router, then I can't really use that tool. Also, we expose these components to designers as a reference, so when new designs are created, they can refer to a single site where they can view each component in isolation and have confidence they are not re-inventing an existing component or adapt a component we already have to make it more general. Is what we are saying is to:

- not test components that use any router related hooks? (barring the cypress tests of course).
- should I re-write the components so they now take values from the likes of `useMatches`, `useLocation` and `useParams` as props?
- not use Storybook for component development, at least those that are going to use router hooks.

That seems quite limiting and would be a barrier to wider adoption. I don't think my manager(s) would agree to that, for instance, many places I've worked, there have been contractual obligations to provide this unit tests and designs. 

Are router hooks especially, really so internal to remix that they cannot be exported to be used in a test/story and if so, would it not be better to  write some custom RTL render extension and storybook extension that can do this? I guess I'm wrong or likely you would have, or be saying, it is coming at some point.

To end on a positive, I've really enjoyed learning remix, I see so many benefits to and conventional fixes to stuff I've had to put together myself over the years.

**Reply by sergiodxa** [reply]
If I where to use Storybook or unit test a component in Remix I would try to make that component not depend on Remix directly Asha receive anything as props as you mentioned on (2)

> Are router hooks especially, really so internal to remix that they cannot be exported to be used in a test/story

The hooks are exposed, and you could mock them probably, the issue is they depend on a context which is not exposed so you can’t mock that, you need to mock the hooks and components directly.

> would it not be better to write some custom RTL render extension

I think the team wants to do this z

**Comment by marksmall** [comment]
Thanks for the response @sergiodxa, all your points make sense. If I can and able to help in any way, let me know.

**Comment by MichaelDeBoey** [comment]
We now have a detailed discussion about the preferred testing strategy: https://github.com/remix-run/remix/discussions/2481
